gpiozero/pins/native.py

```diff
@@ -202,113 +206,86 @@ class NativePin(LocalPin):
-            self._level_offset = self._MEM.GPLEV_OFFSET + (number // 32)
-            self._level_shift = number % 32
-            self._pull_offset = self._MEM.GPPUDCLK_OFFSET + (number // 32)
-            self._pull_shift = number % 32
-            self._edge_offset = self._MEM.GPEDS_OFFSET + (number // 32)
-            self._edge_shift = number % 32
-            self._rising_offset = self._MEM.GPREN_OFFSET + (number // 32)
-            self._rising_shift = number % 32
-            self._falling_offset = self._MEM.GPFEN_OFFSET + (number // 32)
-            self._falling_shift = number % 32
-            self._when_changed = None
-            self._change_thread = None
-            self._change_event = Event()
-            self.function = 'input'
-            self.pull = 'up' if cls.PI_INFO.pulled_up('GPIO%d' % number) else 'floating'
-            self.bounce = None
-            self.edges = 'both'
-            cls._PINS[number] = self
-            return self
-
-    def __repr__(self):
-        return "GPIO%d" % self._number
-
-    @property
-    def number(self):
-        return self._number
+    def __init__(self, factory, number):
+        super(NativePin, self).__init__(factory, number)
+        self._func_offset = self.factory.mem.GPFSEL_OFFSET + (number // 10)
+        self._func_shift = (number % 10) * 3
+        self._set_offset = self.factory.mem.GPSET_OFFSET + (number // 32)
+        self._set_shift = number % 32
+        self._clear_offset = self.factory.mem.GPCLR_OFFSET + (number // 32)
+        self._clear_shift = number % 32
+        self._level_offset = self.factory.mem.GPLEV_OFFSET + (number // 32)
+        self._level_shift = number % 32
+        self._pull_offset = self.factory.mem.GPPUDCLK_OFFSET + (number // 32)
+        self._pull_shift = number % 32
+        self._edge_offset = self.factory.mem.GPEDS_OFFSET + (number // 32)
+        self._edge_shift = number % 32
+        self._rising_offset = self.factory.mem.GPREN_OFFSET + (number // 32)
+        self._rising_shift = number % 32
+        self._falling_offset = self.factory.mem.GPFEN_OFFSET + (number // 32)
+        self._falling_shift = number % 32
+        self._when_changed = None
+        self._change_thread = None
+        self._change_event = Event()
+        self.function = 'input'
+        self.pull = 'up' if factory.pi_info.pulled_up(self.address[-1]) else 'floating'
+        self.bounce = None
+        self.edges = 'both'
 
     def close(self):
+        self.frequency = None
         self.when_changed = None
         self.function = 'input'
-        self.pull = 'up' if self.PI_INFO.pulled_up('GPIO%d' % self.number) else 'floating'
+        self.pull = 'up' if self.factory.pi_info.pulled_up(self.address[-1]) else 'floating'
 
     def _get_function(self):
-        return self.GPIO_FUNCTION_NAMES[(self._MEM[self._func_offset] >> self._func_shift) & 7]
+        return self.GPIO_FUNCTION_NAMES[(self.factory.mem[self._func_offset] >> self._func_shift) & 7]
```

Comment:
Well, the factory can be seen as the thing that owns the GPIO registers - I quite like the symmetry of having a single object that represents that single block of memory, with the pin objects representing operations against subsets of that block.